src/tiled/scriptdialog.cpp

```diff
@@ -0,0 +1,270 @@
+{
+    // make the right-hand column more likely to stretch
+    m_gridLayout->setColumnStretch(0, leftColumnStretch);
+    m_gridLayout->setColumnStretch(1, rightColumnStretch);
+    m_rowIndex = 0;
+
+    addNewRow();
+}
+void ScriptDialog::clear()
+{
+    deleteAllFromLayout(layout());
+    initializeLayout();
+}
+
+QLabel *ScriptDialog::addHeading(const QString &text, bool fillRow)
+{
+    addNewRow();
+    QLabel *label = newLabel(text);
+    m_gridLayout->addWidget(label, m_rowIndex, m_widgetsInRow, 1, fillRow? -1 : 1); // max width
+    label->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
+    m_widgetsInRow++;
+    if (fillRow) {
+        label->setWordWrap(true);
+        addNewRow();
+    }
+    return label;
+}
+QLabel *ScriptDialog::addLabel(const QString &text)
+{
+    moveToRightColumn();
+    return (QLabel *)addDialogWidget(newLabel(text));
+}
+
+QFrame *ScriptDialog::addSeparator(const QString &labelText)
+{
+    addNewRow();
+    m_rowLayout = new QHBoxLayout();
+    // we don't use addDialogWidget() here so that we can make the size of the separator
+    // level independent of the size of the left column.
+    m_gridLayout->addLayout(m_rowLayout, m_rowIndex, 0, 1, -1); // span entire row
+    if (!labelText.isEmpty()) {
+        QLabel *separatorLabel;
+        separatorLabel= newLabel(labelText);
+        separatorLabel->setWordWrap(false);
+        m_rowLayout->addWidget(separatorLabel, 1);
+    }
+    QFrame *line = new QFrame(this);
+    line->setFrameShape(QFrame::HLine);
+    line->setFrameShadow(QFrame::Sunken);
+    m_rowLayout->addWidget(line, rightColumnStretch); // higher stretch
+    addNewRow();
+    return line;
+}
+QLineEdit *ScriptDialog::addTextInput(const QString &labelText, const QString &defaultValue)
+{
+    checkIfSameType("QLineEdit");
+    moveToRightColumn();
+    if (!labelText.isEmpty()) {
+        QLabel *lineEditLabel = newLabel(labelText);
+        addDialogWidget(lineEditLabel);
+    }
+    return (QLineEdit *)addDialogWidget(new QLineEdit(defaultValue, this));
+}
+QDoubleSpinBox *ScriptDialog::addNumberInput(const QString &labelText)
+{
+    checkIfSameType("QDoubleSpinBox");
+    moveToRightColumn();
+    if (!labelText.isEmpty()) {
+        QLabel *numberLabel = newLabel(labelText);
+        addDialogWidget(numberLabel);
+    }
+    return (QDoubleSpinBox *)addDialogWidget(new QDoubleSpinBox(this));
+}
+QSlider *ScriptDialog::addSlider(const QString &labelText)
+{
+    checkIfSameType("QSlider");
+    moveToRightColumn();
+    if (!labelText.isEmpty()) {
+        QLabel *sliderLabel= newLabel(labelText);
+        addDialogWidget(sliderLabel);
+    }
+    QSlider *horizontalSlider = new QSlider(this);
+    horizontalSlider->setOrientation(Qt::Horizontal);
+    return (QSlider *)addDialogWidget(horizontalSlider);
+}
+
+QCheckBox *ScriptDialog::addCheckBox(const QString &labelText, bool defaultValue)
+{
+    checkIfSameType("QCheckBox");
+    moveToRightColumn();
```

Comment:
Actually, could these calls be merged into `addDialogWidget`? We don't need to mention the class name, since `addDialogWidget` could obtain that through `widget->metaObject()->className()`. In fact I guess we don't need to compare class names, but could just compare the `QMetaObject*`.